Suspendable request helpers

gavin-norman-sociomantic opened the pull request.

Commit:
Add helpers encapsulating core logic of suspendable request handling

gavin-norman-sociomantic commented:
Based on #8, #19, #22, #17. Blocked until those PRs are merged.

gavin-norman-sociomantic commented:
Prerequisites merged. Unblocked.

nemanja-boric-sociomantic commented on src/swarm/neo/client/mixins/SuspendableRequestCore.d:
```diff
@@ -0,0 +1,1165 @@
+
+    /// Struct encapsulating the logic for inc/decrementing the request's
+    /// SuspendableRequestSharedWorkingData.ready_for_state_change counter.
+    private struct StateChangeReadinessTracker
```
I like encapsulation of concerns this struct is doing.

gavin-norman-sociomantic commented:
Added release notes.

nemanja-boric-sociomantic commented on src/swarm/neo/client/mixins/SuspendableRequestCore.d:
```diff
@@ -0,0 +1,1165 @@
+
+    public void reset ( )
+    {
+        this.ready_for_state_change = false;
```
`this.all_nodes_initialiser.reset()` for future-proofness (as it is empty right now, might not be in the future)?

nemanja-boric-sociomantic commented on src/swarm/neo/client/mixins/SuspendableRequestCore.d:
```diff
@@ -0,0 +1,1165 @@
+public AllNodesRequest!(Request, Connector, Disconnected,
+    SuspendableRequestInitialiser!(Request, FillPayload, HandleStatusCode),
+    Handler)
+    createSuspendableRequest ( Request, Connector, Disconnected, FillPayload,
```
I have no better suggestion here, but it's very hard to spot the function's name with the current indentation (all these helper functions have the same problem, but this one is the worst).

gavin-norman-sociomantic replied:
Yeah, I'll see if I can come up with anything better.

nemanja-boric-sociomantic commented on src/swarm/neo/client/mixins/SuspendableRequestCore.d:
```diff
@@ -0,0 +1,1165 @@
+
+        /***********************************************************************
+
+            Changes the desired state to that specified. Sets the desired state
```
`to that specified` or `to the specified`?

gavin-norman-sociomantic replied:
"to that" is correct.

nemanja-boric-sociomantic replied:
Thanks! I was unsure.

nemanja-boric-sociomantic commented on src/swarm/neo/client/mixins/SuspendableRequestCore.d:
```diff
@@ -0,0 +1,1165 @@
+            control message flag to the fiber via the return value of suspend().
+
+            If one or more connections are not ready to change state, the
+            control change doesn not occur. A connection is ready to change the
```
typo `doesn not`

nemanja-boric-sociomantic commented on src/swarm/neo/client/mixins/SuspendableRequestCore.d:
```diff
@@ -0,0 +1,1165 @@
+            // the case where they request a new state change in the notifier.
+            state_changed_in_notifier = this.notifyUser(stopped);
+        }
+        while ( !stopped );
```
I have no words how good this is.

gavin-norman-sociomantic replied:
Hmm wait... there's some conflict between this and https://github.com/gavin-norman-sociomantic/swarm-tsunami/blob/2ae0bb629d97d29219d121f5d569a93473a72355/test/neo/client/request/internal/GetAll.d#L413. I'll have to look into this. It looks like the same logic has been duplicated in two place, in slightly different forms.

nemanja-boric-sociomantic replied:
Ah, ok, I haven't got to that point yet.

nemanja-boric-sociomantic commented on src/swarm/neo/client/mixins/SuspendableRequestCore.d:
```diff
@@ -0,0 +1,1165 @@
+
+    ***************************************************************************/
+
+    private bool notifyUser ( bool stopped )
```
I was trying to see how this flag is used, but it seems that it's not used at all (less complexity, which is good).

gavin-norman-sociomantic replied:
Hmm wait... there's some conflict between this and https://github.com/gavin-norman-sociomantic/swarm-tsunami/blob/2ae0bb629d97d29219d121f5d569a93473a72355/test/neo/client/request/internal/GetAll.d#L413. I'll have to look into this. It looks like the same logic has been duplicated in two places, in slightly different forms. (`stopped` is used in the other module.)

nemanja-boric-sociomantic commented on src/swarm/neo/client/mixins/SuspendableRequestCore.d:
```diff
@@ -0,0 +1,1165 @@
+    /***************************************************************************
+
+        The number of handlers that are currently in the process of sending a
+        control message to the node of waiting for an acknowledgement message
```
`or waiting for an acknowledgement`

nemanja-boric-sociomantic reviewed (commented):
comments for the first commit

gavin-norman-sociomantic commented:
Updated, removing doubled logic in `SuspendableRequestControllerFiber` and the client's GetAll implementation (now simplified).

gavin-norman-sociomantic commented:
Updated:
* Call `AllNodesRequestInitialiser.reset` in `SuspendableRequestInitialiser.reset`.
* Fixed typos.
* Improved formatting of long function signatures.

nemanja-boric-sociomantic commented:
Have you forgot to push?

nemanja-boric-sociomantic commented:
(https://github.com/sociomantic-tsunami/swarm/pull/25/commits/5dc65393c47744638a8b2b61def00ed4dca4e5c5#diff-0ad780cddea50703f1972818dba99322R296 still doesn't call `superinitialiser.reset()` for example)

nemanja-boric-sociomantic commented:
Ah no, it's in the wrong (the second) commit.

gavin-norman-sociomantic commented:
>Ah no, it's in the wrong (the second) commit.

Updated.

nemanja-boric-sociomantic commented on src/swarm/neo/client/mixins/SuspendableRequestCore.d:
```diff
@@ -322,10 +322,12 @@ public struct SuspendableRequestInitialiser ( Request, FillPayload,
-    createSuspendableRequestInitialiser ( Request, FillPayload, HandleStatusCode )
-    ( RequestOnConn.EventDispatcherAllNodes conn, Request.Context* context,
-    FillPayload fill_payload, HandleStatusCode handle_status_code )
+public
```
this is still in the wrong commit

gavin-norman-sociomantic replied:
Ah shit, that too.

gavin-norman-sociomantic replied:
Updated. (There was a bunch of other stuff wrongly in the second commit too!)

nemanja-boric-sociomantic commented on src/swarm/neo/client/mixins/SuspendableRequestCore.d:
```diff
@@ -580,7 +584,7 @@ public template SuspendableController ( Request, IController, MessageType )
 
             If one or more connections are not ready to change state, the
-            control change doesn not occur. A connection is ready to change the
+            control change does not occur. A connection is ready to change the
```
ditto for all other stuff in this file I guess.

Commit:
Enhance test GetAll request to be suspend/resume/stoppable

This involves adapting the client API, the request API, the client request
handler, and the node request handler.

The client request handler is adapted to use AllNodesRequestCore and
SuspendableRequestCore.

nemanja-boric-sociomantic commented:
Now I'm excited :-).
